src/ServiceControl/Infrastructure/TimeKeeper.cs

```diff
@@ -33,11 +36,13 @@ public Timer NewTimer(Func<Task<bool>> callback, TimeSpan dueTime, TimeSpan peri
                 {
                     log.Error("Reoccurring timer task failed.", ex);
                 }
-                if (reschedule && timers.ContainsKey(timer))
+
+                var timerSelf = timer[0];
+                if (reschedule && timers.ContainsKey(timerSelf))
```

Comment:
Found on TC 

```
[12:46:25][Step 2/2] 2018-04-12 12:46:03,430 [11120] WARN  JetBrains.Container.Log4netDefaultLoggerInitializer - Unhandled exception in current domain: System.ArgumentNullException: Value cannot be null.
[12:46:25][Step 2/2] Parameter name: key
[12:46:25][Step 2/2]    at System.Collections.Concurrent.ConcurrentDictionary`2.ContainsKey(TKey key)
[12:46:25][Step 2/2]    at ServiceControl.Infrastructure.TimeKeeper.<>c__DisplayClass6_0.<NewTimer>b__0(Object _) in C:\BuildAgent\work\7189a56f9f44affc\src\ServiceControl\Infrastructure\TimeKeeper.cs:line 36
[12:46:25][Step 2/2]    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
[12:46:25][Step 2/2]    at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
[12:46:25][Step 2/2]    at System.Threading.TimerQueueTimer.CallCallback()
[12:46:25][Step 2/2]    at System.Threading.TimerQueueTimer.Fire()
[12:46:25][Step 2/2]    at System.Threading.TimerQueue.FireNextTimers(),    at System.Collections.Concurrent.ConcurrentDictionary`2.ContainsKey(TKey key)
[12:46:25][Step 2/2]    at ServiceControl.Infrastructure.TimeKeeper.<>c__DisplayClass6_0.<NewTimer>b__0(Object _) in C:\BuildAgent\work\7189a56f9f44affc\src\ServiceControl\Infrastructure\TimeKeeper.cs:line 36
[12:46:25][Step 2/2]    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
[12:46:25][Step 2/2]    at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
[12:46:25][Step 2/2]    at System.Threading.TimerQueueTimer.CallCallback()
[12:46:25][Step 2/2]    at System.Threading.TimerQueueTimer.Fire()
[12:46:25][Step 2/2]    at System.Threading.TimerQueue.FireNextTimers(), (null)
```

due to closure capturing it is possible that the timer callback fires and the closure sees the local as being null and then `ContainsKey` throws